code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -15,32 +80,654 @@ impl<T: Config> Pallet<T> {
+			channel_id,
+			coin: PrefixedCoin {
+				amount: Amount::from(amount.amount as u64),
+				denom: PrefixedDenom::from_str(amount.denom.as_ref()).map_err(|_| {
+					<CosmwasmVMError<T>>::Ibc("provided asset is not IBC compatible".to_string())
+				})?,
+			},
+			from: vm.contract_address.clone().into_inner(),
+			timeout: ibc_primitives::Timeout::Offset {
+				timestamp: Err(<CosmwasmVMError<T>>::Ibc(
+					"after timeout will have pub interface".to_string(),
+				))?,
+				height: Err(<CosmwasmVMError<T>>::Ibc(
+					"after timeout will have pub interface".to_string(),
+				))?,
+			},
+			to: IbcSigner::from_str(to_address.as_ref())
+				.map_err(|_| <CosmwasmVMError<T>>::Ibc("bad ".to_string()))?,
+		};
+
+		T::IbcRelayer::handle_message(msg)
+			.map_err(|_err| <CosmwasmVMError<T>>::Ibc("failed to send amount".to_string()))
 	}
 
 	pub(crate) fn do_ibc_send_packet(
-		_vm: &mut CosmwasmVM<T>,
-		_channel_id: String,
-		_data: cosmwasm_minimal_std::Binary,
+		vm: &mut CosmwasmVM<T>,
+		channel_id: String,
+		data: cosmwasm_minimal_std::Binary,
 		_timeout: cosmwasm_minimal_std::ibc::IbcTimeout,
 	) -> Result<(), CosmwasmVMError<T>> {
-		Err(Error::<T>::Unsupported.into())
+		let port_id = PortId::from_str(&Self::do_compute_ibc_contract_port(
+			vm.contract_address.as_ref().clone(),
+		))
+		.expect("address is port; qed");
+		let channel_id = ChannelId::from_str(&channel_id)
+			.map_err(|_| CosmwasmVMError::<T>::Ibc("unsupported channel name".to_string()))?;
+
+		T::IbcRelayer::handle_message(HandlerMessage::SendPacket {
+			data: data.to_vec(),
+			timeout: Err(<CosmwasmVMError<T>>::Ibc(
+				"as soon as IBC will provide public timeout".to_string(),
+			))?,
+			channel_id,
+			port_id,
+		})
+		.map_err(|_| CosmwasmVMError::<T>::Ibc("failed to send packet".to_string()))
 	}
 
 	pub(crate) fn do_ibc_close_channel(
-		_vm: &mut CosmwasmVM<T>,
-		_channel_id: String,
+		vm: &mut CosmwasmVM<T>,
+		channel_id: String,
 	) -> Result<(), CosmwasmVMError<T>> {
-		Err(Error::<T>::Unsupported.into())
+		let channel_id = ChannelId::from_str(channel_id.as_ref())
+			.map_err(|_| <CosmwasmVMError<T>>::Ibc("channel name is not valid".to_string()))?;
+		let address: cosmwasm_minimal_std::Addr = vm.contract_address.clone().into();
+
+		let port_id = PortId::from_str(address.as_str())
+			.expect("all pallet instanced contract addresses are valid port names; qwe");
+
+		T::IbcRelayer::handle_message(HandlerMessage::CloseChannel { channel_id, port_id })
+			.map_err(|_| CosmwasmVMError::<T>::Ibc("failed to close channel".to_string()))
 	}
 
 	pub(crate) fn do_compute_ibc_contract_port(address: AccountIdOf<T>) -> String {
 		format!("wasm.{}", Pallet::<T>::account_to_cosmwasm_addr(address))
 	}
 }
+
+#[derive(Default, RuntimeDebug, Eq, PartialEq, Clone)]
+pub struct Router<T: Config> {
+	_marker: PhantomData<T>,
+}
+
+struct MapBinary(Vec<u8>);
+
+impl AsRef<[u8]> for MapBinary {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
```

Comment:
`as_slice()` works ;)

Reply:
&self.0[..] vs &self.0.as_slice() ? imho it is opinionated. let wait for others

Reply:
yeah, I know! Not really important